dashboard/config/routes.rb

```diff
@@ -1038,5 +1038,7 @@
     # Adds the experiment cookie in the User's browser which allows them to experience offline features
     get '/offline/join_pilot', action: :set_offline_cookie, controller: :offline
     get '/offline-files.json', action: :offline_files, controller: :offline
+
+    get '/get_token', to: 'authenticity_token#get_token'
```

Comment:
How is this secured? Can a person execute a get against this via `curl` to obtain a token? Does this make it easier for CSRF/XSS attacks?

Reply:
Agreed. This is a tricky issue. Would this make our application-wide usage of CSRF less secure?

Reply:
Can we use `cancancan` ability checks to validate these requests?

Reply:
As far as I understand, this shouldn't make our application-wide usage of CSRF any less secure. A person could execute a get against this URL to obtain a token, but they could also make a GET against studio.code.org/courses to get a valid token as well. However, I believe in either case, they would not be able to carry out a CSRF attack (trying to make another user perform a state-changing action) because they would still need access to another user's session cookie, and therefore wouldn't be able to obtain a token associated with that user specifically.

We aren't using `cancancan` here because this endpoint needs to work for signed out sessions as well. For signed out sessions, an attacker could in theory request a new token from this endpoint or the homepage, and make a POST request to any endpoint not protected by sign in, but I believe that doesn't qualify as a CSRF attack because they're not forging a request as another user. There's an open question here of if we even need CSRF protection on non-signed in endpoints, but it does at least prevent attackers from hitting those endpoints as different users.

Reply:
Just in case, seems worth protecting this endpoint with a Gatekeeper flag, and have it return an HTTP status code and response that the client functions understand indicating that the feature has temporarily been disabled.